### ISSUE TITLE
How to design TP plan for `nn.GLU`

### ISSUE THREAD

**Issue by yzhangcs** [pytorch/torchtitan#827]
Hi guys, I'm encountering a challenge in designing TP plans for gated MLP, i.e., [nn.GLU](https://pytorch.org/docs/stable/generated/torch.nn.GLU.html#torch.nn.GLU) with packed weights `w12 = [w1, w2]`, followed by a down proj `w3`

The plan for separated `w1` and `w2` is quite straightforward
```
layer_tp_plan = {
    # by default ColwiseParallel input layouts is replicated
    # and RowwiseParallel output layouts is replicated
    "feed_foward.w1": ColwiseParallel(),
    "feed_forward.w2": RowwiseParallel(),
    "feed_forward.w3": ColwiseParallel(),
}
```
However, I'm unsure how to approach this when using packed weights (`w12 = [w1, w2]`) to leverage the fused GLU for better performance.

Could anyone provide some guidance on how to design an effective TP plan for this scenario?
Thank you 

@tianyu-l

**Comment by tianyu-l** [comment]
could you please show a more complete code snippet containing how you would like to use `nn.GLU`? e.g. the FFN layer's `forward`

**Comment by yzhangcs** [comment]
@tianyu-l The typical Llama Gated MLP is defined as, 
`y = self.w3(F.silu(self.w1(x)) * self.w2(x))`
we can easily design tp plans by applying `ColwiseParallel` to `w1` & `w2` first and then `RowwiseParallel` to `w3`

However things could be different once `w1` and `w2` being packed into `w12 = [w1, w2]`, so that
`y = self.w3(F.glu(self.w12(x)))`

I'm not sure how to better arrange parallelisms for `w12` followed by `w3`

**Comment by tianyu-l** [comment]
Can you just apply `ColwiseParallel` to `w12` and `RowwiseParallel` to `w3`?
Note that although it's called `Colwise`, the tensor in `w12` will be split along the 0 dimension, per https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/parallel/style.py#L125

Also `F.glu` accepts a `dim` arg, so you get some flexibility on stacking w1/w2 and their outputs horizontally/vertically anyways.

**Comment by yzhangcs** [comment]
@tianyu-l Thank you! I will check that.

**Comment by yzhangcs** [comment]
mentioned by this comment https://github.com/pytorch/pytorch/issues/88838#issuecomment-2081838245
